### PR TITLE
Fix #1921: Support figure substitutions by locale

### PR DESCRIPTION
**Pull request by tk0miya**
For #1921.
Improve i18n and support figure substitutions by locale (`figure_language`).

### CODE REVIEW

**Comment by tk0miya** [comment]
@shimizukawa @lehmannro any comments or objections?
I would like to merge this before 1.4a1 (in this weekend).

File: sphinx/environment.py [hunk 1]
```diff
@@ -961,6 +962,35 @@ def process_downloads(self, docname, doctree):
...
+                elif filename.lower().endswith('.svg'):
+                    if 'image/svg+xml' not in candidates:
+                        globbed.setdefault('image/svg+xml', []).append(new_imgpath)
+                else:
+                    try:
+                        f = open(filename, 'rb')
+                        try:
+                            imgtype = imghdr.what(f)
+                        finally:
+                            f.close()
```
Comment (shimizukawa) [inline sphinx/environment.py]: Please use `with open` instead of open-try-finally-close.

Comment (shimizukawa) [inline sphinx/environment.py]: (Ah, it's not changed from the original..)

File: sphinx/environment.py [hunk 2]
```diff
@@ -961,6 +962,35 @@ def process_downloads(self, docname, doctree):
...
+                    except (OSError, IOError) as err:
+                        self.warn_node('image file %s not readable: %s' %
+                                       (filename, err), node)
+                    if imgtype and 'image/' + imgtype not in candidates:
+                        globbed.setdefault('image/' + imgtype, []).append(new_imgpath)
+            for key, files in iteritems(globbed):
+                if len(files) == 1:
+                    candidates[key] = files[0]
+                else:
+                    candidates[key] = sorted(files, key=len)[0]  # select by similarity
```
Comment (shimizukawa) [inline sphinx/environment.py]: I think this if-else is not needed. You can just use else clause block even if `len(files) == 1`.

**Comment by shimizukawa** [comment]
LGTM with nits.

**Comment by tk0miya** [comment]
@shimizukawa Thank you for reviewing. I updated the pull request.

**Comment by lehmannro** [comment]
Why is there a separate `figure_language`, and what's the difference to the existing `language` setting?

File: sphinx/config.py [hunk 3]
```diff
@@ -51,6 +51,8 @@ class Config(object):
         # the real default is locale-dependent
         today_fmt = (None, 'env', [str]),
 
+        figure_language = (None, 'env', [str]),
+        figure_language_filename = ('%(root)s.%(language)s%(ext)s', 'env', [str]),
```
Comment (lehmannro) [inline sphinx/config.py]: This does not need the `[str]` type hint, because its default is of type string as well.

Also, %-based string interpolation looks a bit obtuse here.  Maybe switching for `str.format` makes this a little nicer?

Now that I think about it, how would I implement looking into different directories with this (say, `_locale/fr/images/myimage.png`)?

File: sphinx/environment.py [hunk 4]
```diff
@@ -961,6 +962,29 @@ def process_downloads(self, docname, doctree):
...
     def process_images(self, docname, doctree):
         """Process and rewrite image URIs."""
+        def collect_candidates(imgpath, candidates):
+            globbed = {}
+            for filename in glob(imgpath):
+                new_imgpath = relative_path(path.join(self.srcdir, 'dummy'),
+                                            filename)
+                if filename.lower().endswith('.pdf'):
+                    if 'application/pdf' not in candidates:
+                        globbed.setdefault('application/pdf', []).append(new_imgpath)
```
Comment (lehmannro) [inline sphinx/environment.py]: I might be misunderstanding the logic here, but you're trying to stop collecting further candidates for a mimetype if we already have one?  I'd suggest doing that in the loop further below where you actually fill up `candidates`, rather than sprinkling it over here.

File: sphinx/environment.py [hunk 5]
```diff
@@ -961,6 +962,29 @@ def process_downloads(self, docname, doctree):
 
     def process_images(self, docname, doctree):
         """Process and rewrite image URIs."""
+        def collect_candidates(imgpath, candidates):
+            globbed = {}
+            for filename in glob(imgpath):
+                new_imgpath = relative_path(path.join(self.srcdir, 'dummy'),
+                                            filename)
+                if filename.lower().endswith('.pdf'):
```
Comment (lehmannro) [inline sphinx/environment.py]: Also, while we're here, the whole _file extension to mime type_ mapping could be refactored into a mapping, instead of a collection of ifs.

File: sphinx/environment.py [hunk 6]
```diff
@@ -976,25 +1000,13 @@ def process_images(self, docname, doctree):
...
-                                           (filename, err), node)
-                        if imgtype:
-                            candidates['image/' + imgtype] = new_imgpath
+                if self.config.figure_language:
+                    # Search language-specific figures at first
+                    i18n_imguri = get_image_filename_for_locale(imguri, self, force=True)
+                    _, full_i18n_imgpath = self.relfn2path(i18n_imguri, docname)
+                    collect_candidates(full_i18n_imgpath, candidates)
+
+                collect_candidates(full_imgpath, candidates)
```
Comment (lehmannro) [inline sphinx/environment.py:931]: I'm not sure if it's worth the effort, but could we add a short circuit here?  Something along the lines of _if the preferred mime type is already found, do not search again for unlocalized files_?

File: sphinx/util/i18n.py [hunk 7]
```diff
@@ -118,3 +118,19 @@ def find_catalog_source_files(locale_dirs, locale, domains=None, gettext_compact
                     catalogs.add(cat)
 
     return catalogs
+
+
+def get_image_filename_for_locale(filename, env, force=False):
```
Comment (lehmannro) [inline sphinx/util/i18n.py]: This whole method feels weirdly overloaded to me.

It pretends to work off of a locale, but actually operates on the build environment.  It sometimes does name manipulation, sometimes also checks if a file exists.

File: sphinx/ext/graphviz.py [hunk 8]
```diff
@@ -77,7 +78,8 @@ def run(self):
                     'Graphviz directive cannot have both content and '
                     'a filename argument', line=self.lineno)]
             env = self.state.document.settings.env
-            rel_filename, filename = env.relfn2path(self.arguments[0])
+            argument = get_image_filename_for_locale(self.arguments[0], env)
```
Comment (lehmannro) [inline sphinx/ext/graphviz.py]: What happens here if I didn't have a suffix for my Graphviz filename?  I believe this might previously have been supported, while it's now a weird edge case.

File: tests/test_ext_graphviz.py [hunk 9]
```diff
@@ -43,3 +46,15 @@ def test_graphviz_latex(app, status, warning):
 
     macro = 'Hello \\\\includegraphics{graphviz-\w+.pdf} graphviz world'
     assert re.search(macro, content, re.S)
+
+
+@with_app('html', testroot='ext-graphviz', confoverrides={'figure_language': 'xx'})
+def test_graphviz_i18n(app, status, warning):
+    app.builder.build_all()
+    if "dot command 'dot' cannot be run" in warning.getvalue():
```
Comment (lehmannro) [inline tests/test_ext_graphviz.py]: Seeing this is now the third occurrence of `if "dot command...` in this file, could you refactor this?

File: tests/test_intl.py [hunk 10]
```diff
@@ -794,3 +796,101 @@ def test_references(app, status, warning):
     warnings = warning.getvalue().replace(os.sep, '/')
     warning_expr = u'refs.txt:\\d+: ERROR: Unknown target name:'
     yield assert_count(warning_expr, warnings, 0)
+
+
+@with_app(buildername='dummy', testroot='image-glob', confoverrides={'figure_language': 'xx'})
+def test_image_glob_intl(app, status, warning):
```
Comment (lehmannro) [inline tests/test_intl.py:802]: This method is a sequence of the same three instructions over and over again (assert isinstance, assert candidates, assert uri).  Could you please refactor this?

File: sphinx/transforms.py [hunk 11]
```diff
@@ -556,6 +556,27 @@ def get_ref_key(node):
                 node['entries'] = new_entries
 
 
+class FigureTranslator(Transform):
+    """
+    Substitute language-specific figures for original one.  This transform looks
+    for translated images by filename.  The filename that is being looked for is
+    constructed as follow :samp:`filename.{language}.ext`.  For example, for a
```
Comment (lehmannro) [inline sphinx/transforms.py]: While the sample is nice in illustrating what happens here, it actually depends on the specific configuration value, no?

File: sphinx/transforms.py [hunk 12]
```diff
@@ -556,6 +556,27 @@ def get_ref_key(node):
...
+    filename ``picture.png``, when document is generated for the french language,
```
Comment (lehmannro) [inline sphinx/transforms.py]: Grammar nit:  Languages are generally capitalized in English.

File: sphinx/config.py (hunk 3 above)
Comment (tk0miya) [inline sphinx/config.py]: > This does not need the [str] type hint, because its default is of type string as well.

I didn't know about the default. But explicit type hinting is better for me.
(it looks like other variables also uses `[str]` hinting)

> Also, %-based string interpolation looks a bit obtuse here. Maybe switching for str.format makes this a little nicer?

And, I changed the `figure_lanuage_filename` to `str.format` style now.

> Now that I think about it, how would I implement looking into different directories with this (say, _locale/fr/images/myimage.png)?

This replaces relative path from docname, not document root.
So it is difficult to refer shared image directories, now.

File: sphinx/environment.py (hunk 4 above)
Comment (tk0miya) [inline sphinx/environment.py]: No, it continues collecting.
This pull request tries to search candidates twice (first is for `figure_language`, second is for original one). Line 971 is used for the second collecting to stop overriding if translated image already found.

File: sphinx/environment.py (hunk 5 above)
Comment (tk0miya) [inline sphinx/environment.py]: Indeed. It comes from original code, but I'll refactor it lator.

File: sphinx/environment.py (hunk 6 above)
Comment (tk0miya) [inline sphinx/environment.py:931]: This process is executed on reading phase.
So, at this moment, any preferred mime types could not be determined yet, I think.
Are there any short circuits?

**Comment by tk0miya** [comment]
> Why is there a separate figure_language, and what's the difference to the existing language setting?

I thought anyone requested it, but my memory is wrong. so no special reason.

File: sphinx/config.py (hunk 3 above)
Comment (lehmannro) [inline sphinx/config.py]: > I didn't know about the default. But explicit type hinting is better for me.
> (it looks like other variables also uses [str] hinting)

Yes, but only when the default is not of type string.  The default value already contributes to the list of types a configuration value can take, so adding it here adds _no_ value.

> This replaces relative path from docname, not document root.
> So it is difficult to refer shared image directories, now.

Ah, well, might leave it to another PR to actually do this then.

File: sphinx/environment.py (hunk 4 above)
Comment (lehmannro) [inline sphinx/environment.py]: Well, I don't think we disagree then?
All the `if X not in candidates: globbed.setdefault(X, []).append(new_imgpath)` can be replaced with a simple `globbed.setdefault…` and further down, you'd have:

``` python
for key, files in iteritems(globbed):
  if key not in candidates:
    candidates[key] = sorted(files, key=len)[0]  # select by similarity
```

File: sphinx/environment.py (hunk 6 above)
Comment (lehmannro) [inline sphinx/environment.py:931]: Hm, right!

File: sphinx/util/i18n.py (hunk 7 above)
Comment (lehmannro) [inline sphinx/util/i18n.py]: My proposal would be two functions to make the intent clear, one which has the fallback logic and one which just pure, dumb name translation.

**Comment by lehmannro** [comment]
> > Why is there a separate figure_language, and what's the difference to the existing language setting?
> 
> I thought anyone requested it, but my memory is wrong. so no special reason.

I would recommend _at least_ setting it to `language` by default then.  I cannot think of any use case where you'd want to translate your images to _another_ language than your text.  Maybe this is better suited as a _boolean_ setting to enable/disable image filename translations?

File: sphinx/environment.py (hunk 4 above)
Comment (tk0miya) [inline sphinx/environment.py]: Oh, sorry. I misread your comments. I agree with you.
But, it is already refactored by other comment like following:

```
        def collect_candidates(imgpath, candidates):
            globbed = {}
            for filename in glob(imgpath):
                rel_imgpath = relative_path(path.join(self.srcdir, 'dummy'),
                                            filename)
                mimetype = guess_mimetype(filename)
                if mimetype and mimetype not in candidates:
                    globbed.setdefault(mimetype, []).append(rel_imgpath)
            for key, files in iteritems(globbed):
                candidates[key] = sorted(files, key=len)[0]  # select by similarity
```

**Comment by tk0miya** [comment]
Ok, I will merge `figure_language` into `language` tomorrow. It came from my misunderstanding.

**Comment by tk0miya** [comment]
Note:
I made small commits for each review comments now.
I will squash them before merging.

**Comment by lehmannro** [comment]
Thanks, I'll take a final look after your changes then! :-)

On squashing, I see this relatively often here in Sphinx, but it's generally frowned upon in other projects.  I think the main advantage of /not squashing/ commits is that others can actually follow how a certain feature evolved.  If they're only interested in actual features, they can look at merges-only.  Opinions or policies I've missed?

File: sphinx/config.py (hunk 3 above)
Comment (tk0miya) [inline sphinx/config.py]: Ok. I read the `Config.check_types` and I understand type hinting is no effect.
But, it is still **worth** to me. The config variable declarations having type hinting gives type-information **intentionally** to code readers. (maybe it is useful for document generation :-)
If none, of course it works fine, but code readers might not notice the type-information. because it is implicit.

Again, I think explicit type hinting is better than implicit one even if there is no effect.
What do you think?
If we disagree about this, I will remove these declarations.

File: sphinx/util/i18n.py (hunk 7 above)
Comment (tk0miya) [inline sphinx/util/i18n.py]: OK, I will split it into for searching and for building filename.

File: sphinx/ext/graphviz.py (hunk 8 above)
Comment (tk0miya) [inline sphinx/ext/graphviz.py]: Try to substitute to `filename.{language}` (by default `figure_language_filename`).
We still keep on supporting them.

**Comment by tk0miya** [comment]
@lehmannro I finished refactoring. Could you review this again, please?

In this case, I want to squash my commits to five or six commits. I stacked fixes step by step for every review comment. So, there are some meaningless commits.

**Comment by tk0miya** [comment]
ping, any comments?

**Comment by shimizukawa** [comment]
LGTM (again)!
I think now time to merge this :smiley:

**Comment by tk0miya** [comment]
Thank you for reviewing. I just merged!
